[WebAuthn] CBOR encoded extensions not passed along during assertions

pascoej opened the pull request:
#### c93cca1a0e69b96291e37540a6b2e69468d09760
<pre>
[WebAuthn] CBOR encoded extensions not passed along during assertions
<a href="https://bugs.webkit.org/show_bug.cgi?id=242913">https://bugs.webkit.org/show_bug.cgi?id=242913</a>
rdar://96912101

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
* Source/WTF/wtf/cocoa/SpanCocoa.h:
(WTF::asUInt8Span):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::fromCBOR):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel convertToCoreCreationOptionsWithOptions:]):
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
Pass along CBOR encoded extension to ASC, use span to avoid copy. Rest of callsites to be fixed in
<a href="https://bugs.webkit.org/show_bug.cgi?id=242919.">https://bugs.webkit.org/show_bug.cgi?id=242919.</a>

Canonical link: <a href="https://commits.webkit.org/252626@main">https://commits.webkit.org/252626@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1017,7 +1017,10 @@ - (void)makeCredentialWithClientDataHash:(NSData *)clientDataHash options:(_WKPu
         result.authenticatorAttachmentString = toString(*attachment);
-    result.extensions = authenticationExtensionsClientInputs(options.extensions);
+    if (options.extensionsCBOR)
+        result.extensions = WebCore::AuthenticationExtensionsClientInputs::fromCBOR(vectorFromNSData(options.extensionsCBOR));
```
It is a bit unfortunate to copy the bytes from a NSData to a Vector, just to decode it.

Ideally, AuthenticationExtensionsClientInputs::fromCBOR() would take in a `Span<const uint8_t>` so we could pass in data without copying it. It is cheap to construct a Span from a Vector or a NSData because it doesn't involve copying.

pascoej replied:
Ooh, that's really good to know. Updating CBORReader with this.

brentfulgham replied:
We should check (as a follow-up patch) if there are other cases where we lack this optimization.

pascoej replied:
https://bugs.webkit.org/show_bug.cgi?id=242919

cdumez commented on Source/WTF/wtf/cocoa/SpanCocoa.h:
```diff
@@ -41,6 +41,12 @@ inline Span<const std::byte> asBytes(const RetainPtr<NSData>& data)
     return asBytes(data.get());
 }
 
+inline Span<const uint8_t> spanFromNSData(NSData* data)
```
Any chance we could use asBytes() from SpanCocoa.h?

pascoej replied:
I tried this, converting everything to use std::byte instead of uint8_t turns out to be pretty involved as` Span<const uint8_t>` and `Span<const std::byte>` don't seem to be easily interchangeable.

cdumez replied:
Ok, maybe we can name it `asUInt8Span()` to be a little more consistent with `asBytes()`.

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
```diff
@@ -33,7 +33,7 @@
-std::optional<AuthenticationExtensionsClientInputs> AuthenticationExtensionsClientInputs::fromCBOR(const Vector<uint8_t>& buffer)
+std::optional<AuthenticationExtensionsClientInputs> AuthenticationExtensionsClientInputs::fromCBOR(Span<const uint8_t> buffer)
 {
     std::optional<cbor::CBORValue> decodedValue = cbor::CBORReader::read(buffer);
```
I believe you're constructing a Vector from the Span here and thus copying the data still.

cbor::CBORReader::read() should take a Span too.

cdumez reviewed (approved):
Looks good.

pascoej committed:
[WebAuthn] CBOR encoded extensions not passed along during assertions
https://bugs.webkit.org/show_bug.cgi?id=242913
rdar://96912101

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
* Source/WTF/wtf/cocoa/SpanCocoa.h:
(WTF::asUInt8Span):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::fromCBOR):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(+[_WKWebAuthenticationPanel convertToCoreCreationOptionsWithOptions:]):
(+[_WKWebAuthenticationPanel convertToCoreRequestOptionsWithOptions:]):
Pass along CBOR encoded extension to ASC, use span to avoid copy. Rest of callsites to be fixed in
https://bugs.webkit.org/show_bug.cgi?id=242919.

Canonical link: https://commits.webkit.org/252626@main
